docs/changelog.rst

```diff
@@ -109,6 +109,9 @@ Build system improvements:
 
 * Fix a harmless warning from CMake 3.21 with the classic Python discovery.
   `#3220 <https://github.com/pybind/pybind11/pull/3220>`_
+  
+* Eigen repo and version can now be selected, and reconfigures faster.
```

Comment:
"reconfigures" or "reconfigured"? (I cannot tell)

Reply:
Does the std::free deserve a mention?

Reply:
I didn't think so (just took off the needs-changlog), but if you think it does, we can?

Reply:
On the fence, it does fix a rare bug, but it's also a really minor change.

Reply:
For free: I think not mentioning is fine.

For eigen, how about simply:

Eigen repo and version can now be specified as cmake options.

Rationale: Mentioning "cmake" is important to establish relevant context. It's obvious that it's then easier to reconfigure (it doesn't have to be stated).

Reply:
Before, it had to reach out to the Internet every time you reconfigured to see if the version (3.3.8) was still the thing you had locally. Now, we use the hash instead, so it knows that the hash is unchanged, and doesn't reach out to the internet to look at the repository for possible tag updates.

I don't think it affects anyone that much, though, as it is tests only, so we can leave off mentioning the improvement.

Reply:
I'm fine to leave off rare/small bugfixes, to highlight the bigger ones. The really tiny ones can be seen in the git logs.

Reply:
> Before, it had to reach out to the Internet every time you reconfigured to see if the version (3.3.8) was still the thing you had locally. Now, we use the hash instead, so it knows that the hash is unchanged, and doesn't reach out to the internet to look at the repository for possible tag updates.

I was wondering but that totally wasn't clear to me. I think it's worth spelling that out in the changelog (~copy-paste from here).

Reply:
I don't think we have that many users building our tests with DOWNLOAD_EIGEN=ON, so I think it's fine as is (you have to have a pretty recent CMake to even do that). Too much text would highlight this change as more important than it is.